src/vulkan/Context.zig

```diff
@@ -72,13 +74,124 @@ pub fn init(
     };
 
     const instance_handle = try self.vkb.createInstance(&instance_create_info, null);
-    self.vki = try InstanceDispatch.load(instance_handle, self.vkb.dispatch.vkGetInstanceProcAddr);
-    self.instance = Instance.init(instance_handle, &self.vki);
+    const vki = try InstanceDispatch.load(instance_handle, self.vkb.dispatch.vkGetInstanceProcAddr);
+    self.instance = Instance.init(instance_handle, &vki);
+
+    var surface: vk.SurfaceKHR = undefined;
+    if (glfw.createWindowSurface(instance_handle, window, null, &surface) != 0)
+        return error.SurfaceLostKHR;
 
+    self.physical_device = try pickPhysicalDevice(self, surface, allocator);
     return self;
 }
 
 pub fn deinit(self: *Self) void {
     self.instance.destroyInstance(null);
     self.allocator.destroy(self);
 }
+
+// scores the devices based on supported extensions, surfaces and queues
+fn pickPhysicalDevice(self: *Self, surface: vk.SurfaceKHR, allocator: Allocator) !vk.PhysicalDevice {
+    const available_devices = try self.instance.enumeratePhysicalDevicesAlloc(allocator);
+    defer allocator.free(available_devices);
+
+    // score cant be less then 0 anyway
+    var max_score: u64 = 0;
+    var max_device: vk.PhysicalDevice = undefined;
+
+    for (available_devices) |pdev| {
+        const extension_score = checkExtensionSupport(self, pdev, allocator) catch continue;
+        const surface_score = checkSurfaceSupport(self, pdev, surface, allocator) catch continue;
+        const queue_score = checkDeviceQueueSupport(self, pdev, surface, allocator) catch continue;
+        if (max_score < extension_score + surface_score + queue_score) {
```

Comment:
I'd store `extension_score + surface_score + queue_score` in a variable rather than typing it twice